# Decision Tree Regression Project  
# Medical Cost Prediction & Feature Impact Analysis

---

## **Introduction**

This project focuses on building a deep understanding of how machine learning models can be used to model real world problems, specifically in the healthcare domain. Using the Medical Cost Personal Dataset, the goal is not only to predict insurance charges but also to analyze how different features influence model behavior and performance.

Rather than treating this as a simple regression task, this project is approached as a full machine learning experiment, where different modeling choices, preprocessing techniques, and assumptions are carefully tested and evaluated.

---

## **Project Aim**

The aim of this project is to develop, analyze, and compare multiple regression models to accurately predict medical insurance costs while gaining a strong conceptual and practical understanding of machine learning techniques.

---

## **Objectives**

1. Build a solid understanding of regression algorithms through hands-on implementation
2. Perform in-depth exploratory data analysis (EDA) to uncover patterns and relationships
3. Analyze the effect of key features (especially BMI and smoking) on medical costs
4. Apply and compare different preprocessing techniques:
    - Encoding categorical variables
    - Feature scaling
    - Log transformation (for skewed data)
5. Train and evaluate multiple models, such as:
    - Linear Regression
    - Polynomial Regression
    - Support Vector Regression (SVR)
    - Decision Tree
    - Random Forest
6. Perform model comparison and critical analysis, not just reporting scores
7. Understand why a model performs well or fails, not just the final result

---
## **Dataset Description**

The dataset includes the following features:

- age → Age of the individual
- sex → Gender
- bmi → Body Mass Index
- children → Number of dependents
- smoker → Smoking status
- region → Residential area
- charges → Medical insurance cost (target variable)

The dataset used in this project is available on Kaggle: [Dataset Link](https://www.kaggle.com/datasets/d3lhomi10/medical-cost-personal-dataset)

---

## **Problem Type**

This is a Supervised Machine Learning – Regression Problem, where the goal is to predict a continuous target variable (charges).

---

Project Approach

This project is structured as a complete ML pipeline, including:

1. Data Exploration (EDA)
2. Data Preprocessing
3. Feature Engineering
4. Model Training
5. Model Evaluation
6. Model Comparison
7. Insights & Conclusions

The focus is not only on building a working model, but on understanding each step deeply and justifying every decision.




# **Importing the Libraries**

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# **Data preprocessing**

## **Data Cleaning Check Template**
This template is designed to quickly assess the quality of any dataset before building machine learning models or performing analysis.

It provides a structured overview of the dataset by checking for common data issues such as:

- Missing values

- Duplicate rows

- Incorrect data types

- Outliers

- Distribution of numerical features

- Categorical feature consistency

**What This Template Does**

- Displays basic dataset information (shape, data types)

- Identifies missing values and duplicates

- Summarizes numerical and categorical features

- Detects potential outliers using the IQR method

- Highlights columns with low unique values for quick inspection

How to Use

1. Load your dataset using Pandas  

2. Call the function:

In [2]:
def data_quality_report(df):

    print("DATA QUALITY REPORT")
    
    # Print a separator line for better readability
    
    print("=" * 50)
    print("BASIC INFO")
    print("=" * 50)
    
    # Show general information about the dataset (columns, data types, non-null values)
    print(df.info())
    
    # Show number of rows and columns
    print("\n" + "=" * 50)
    print("SHAPE OF DATA")
    print("=" * 50)
    print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
    
    # Check for missing (null) values in each column
    print("\n" + "=" * 50)
    print("MISSING VALUES")
    print("=" * 50)
    missing = df.isnull().sum()
    
    # Display only columns that have missing values
    print(missing[missing > 0])
    
    # Check for duplicate rows
    print("\n" + "=" * 50)
    print("DUPLICATES")
    print("=" * 50)
    print(f"Duplicate rows: {df.duplicated().sum()}")
    
    # Display data types of each column
    print("\n" + "=" * 50)
    print("DATA TYPES")
    print("=" * 50)
    print(df.dtypes)
    
    # Summary statistics for numerical columns (mean, std, min, max, etc.)
    print("\n" + "=" * 50)
    print("NUMERICAL SUMMARY")
    print("=" * 50)
    print(df.describe())
    
    # Summary for categorical (object) columns
    print("\n" + "=" * 50)
    print("CATEGORICAL SUMMARY")
    print("=" * 50)
    print(df.describe(include=['object']))
    
    # Show unique values for columns with low number of distinct values
    # Useful for detecting categories, errors, or inconsistencies
    print("\n" + "=" * 50)
    print("UNIQUE VALUES (LOW CARDINALITY)")
    print("=" * 50)
    for col in df.columns:
        if df[col].nunique() < 10:  # Only show columns with few unique values
            print(f"{col}: {df[col].unique()}")
            
    # correlation
    print("\n" + "=" * 50)
    print("CORRELATION MATRIX")
    print("=" * 50)
    print(df.corr(numeric_only=True))
    
    # Detect outliers using the IQR (Interquartile Range) method
    print("\n" + "=" * 50)
    print("OUTLIERS CHECK (IQR METHOD)")
    print("=" * 50)
    
    # Loop through only numerical columns
    for col in df.select_dtypes(include=np.number).columns:
        Q1 = df[col].quantile(0.25)  # 25th percentile
        Q3 = df[col].quantile(0.75)  # 75th percentile
        IQR = Q3 - Q1  # Interquartile range
        
        # Count rows that fall outside the normal range
        outliers = df[(df[col] < Q1 - 1.5 * IQR) | (df[col] > Q3 + 1.5 * IQR)]
        print(f"{col}: {len(outliers)} outliers")

# **Load dataset**
Apply Data Cleaning Check Template

In [ ]:
dataset = pd.read_csv("insurance.csv")
data_quality_report(dataset)

DATA QUALITY REPORT
BASIC INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB
None

SHAPE OF DATA
Rows: 1338, Columns: 7

MISSING VALUES
Series([], dtype: int64)

DUPLICATES
Duplicate rows: 1

DATA TYPES
age           int64
sex          object
bmi         float64
children      int64
smoker       object
region       object
charges     float64
dtype: object

NUMERICAL SUMMARY
               age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397 

# **Cleaning the data** 

### remove the duplication

In [4]:
# Step 2: Remove duplicates FROM the cleaned dataset
dataset_clean = dataset.drop_duplicates().copy()

# Final result
print(dataset_clean)

      age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1337 rows x 7 columns]


# **Select Independent Variable (X) and Dependent Variable (y)**

Note: The concepts of Independent and Dependent variables were discussed earlier.

If you would like to review or gain a clearer understanding, please refer to this section: [Go to Independent variable VS Dependent variable Section](https://github.com/Hazem1695/ml-concept-briefs)

In [5]:
# Separate independent variable(features (X)) by selecting all rows and all columns except the last one
X = dataset_clean.iloc[:, :-1].values
# Separate the dependent variable (Target(y)) by selecting the last column
y = dataset_clean.iloc[:, -1].values

# **Encoding categorical data**

## Encoding the Indpendent Variable Using One Hot Encoding

Note: I explained One-Hot Encoding earlier.

If you missed it or want a deeper understanding of how it works, you can find it here: [Go to One-Hot Encoding Section](https://github.com/Hazem1695/ml-concept-briefs)

In [6]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
# One-hot encode the 'color' column (index 2) while keeping other features
# Initialize ColumnTransformer to handle categorical data
# [2] targets the 'color' column (index 2)
ct = ColumnTransformer(transformers = [('encoder', OneHotEncoder(sparse_output=False),[1,4,5])], remainder = 'passthrough')
# Execute transformation and convert to NumPy ndarray for model training
X = np.array(ct.fit_transform(X))

# **Splitting the dataset into the Train set and Test set**

In [7]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train , y_test = train_test_split(X, y , test_size = 0.2, random_state = 0)

# **Decision Tree Regression Model**

## Training the Building Decision Tree Regression model on the Training set

In [8]:
from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor(max_depth=6, min_samples_split=5, min_samples_leaf=3, random_state = 0)
regressor.fit(X_train, y_train)

DecisionTreeRegressor(max_depth=6, min_samples_leaf=3, min_samples_split=5,
                      random_state=0)

# **Preticting the Test set results**

In [9]:
y_pred = regressor.predict(X_test)

np.set_printoptions(precision=2)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[ 3969.29  1633.96]
 [10681.69  8547.69]
 [10681.69  9290.14]
 [34138.35 32548.34]
 [ 8809.78  9644.25]
 [ 5250.06  2680.95]
 [ 2210.65  2198.19]
 [ 1683.81  1241.57]
 [ 2210.65  2710.83]
 [12111.02 12235.84]
 [10681.69  8280.62]
 [19418.22 17043.34]
 [14062.01 13974.46]
 [10681.69  8219.2 ]
 [ 5809.8   5472.45]
 [ 5809.8   2438.06]
 [ 5595.45  5267.82]
 [18742.54  3490.55]
 [ 8654.06  6640.54]
 [14062.01 14692.67]
 [ 3969.29  1622.19]
 [13516.08 13224.69]
 [ 3969.29  1256.3 ]
 [ 2151.97  2643.27]
 [ 3969.29  1674.63]
 [ 8654.06  4667.61]
 [ 5250.06  3732.63]
 [10681.69 11552.9 ]
 [ 3968.42  3756.62]
 [38113.55 37465.34]
 [ 7795.81  8059.68]
 [53237.95 47462.89]
 [12111.02 10577.09]
 [12111.02 20630.28]
 [17305.79 14571.89]
 [17885.97 36580.28]
 [ 8771.63  8347.16]
 [36560.35 51194.56]
 [ 8809.78  8428.07]
 [ 3969.29  1880.49]
 [34138.35 33475.82]
 [ 3968.42  2867.12]
 [18742.54  4564.19]
 [47955.67 47496.49]
 [36560.35 36149.48]
 [ 8771.63  8125.78]
 [10681.69 19749.38]
 [ 7795.81  7

# **Evaluating the Model Performance**

In [10]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.8417287461053684

In [11]:
y_trained = regressor.predict(X_train)
r2_score(y_train, y_trained)

0.8978449133573384

# **Model Comparison Table**

| Configuration              | Train R² | Test R²           |
| -------------------------- | -------- | ----------------- |
| Tuned, no log, max_depth=3 | 0.8616   | 0.8306            |
| No tune, no log            | 0.9983   | 0.6860            |
| No tune, with log          | 0.9945   | -385,763,830.97  |
| Tuned, with log            | 0.8014   | -340,650,637.32  |
| Tuned, no log, max_depth=5 | 0.8828   | 0.8410            |
| Tuned, no log, max_depth=7 | 0.9061   | 0.8277            |
| **Tuned, no log, max_depth=6** | **0.8978**   | **0.8417**      |

---

# **Important Insights**

##### **1. Model Performance is Misleading Without Generalization**

The Decision Tree Regressor achieved an extremely high training score (0.99) without tuning, but the test performance dropped significantly. This shows that relying only on training performance can lead to wrong conclusions, as the model was clearly overfitting and failing to generalize.

##### **2. Model Complexity Must Be Controlled**

Adjusting the tree depth had a direct impact on performance. Increasing depth improved results up to a certain point, but after max_depth=6, the model started to overfit. This demonstrates that controlling model complexity is essential to achieve a good balance between learning patterns and avoiding noise.

##### **3. Preprocessing is Not One Size Fits All**

Applying log transformation to the target variable resulted in extremely poor performance. This highlights that preprocessing techniques are not universally beneficial and must be chosen based on the model and data characteristics. Incorrect preprocessing can significantly harm model performance.

##### **4. Small Changes Can Lead to Big Differences**

The difference between `max_depth=5`, `6`, and `7` shows how small parameter changes can noticeably affect performance. A slight increase in complexity improved the model, but going further led to degradation, emphasizing the importance of careful experimentation.

---

# **Conclusion**

This experiment supports a critical principle in machine learning: a model should not be judged based on a single configuration or preprocessing approach. The same model can perform poorly or strongly depending on how the data is handled and how its parameters are tuned. The best result in this case was achieved after adjusting the model complexity and avoiding unsuitable preprocessing, demonstrating that performance improvements often come from better data handling rather than changing the model itself.

Testing different configurations and preprocessing techniques is essential to understand model behavior and achieve the best performance.

---

# **Model Comparison**

| Model Type                | Configuration                   | R² Score (Test) |
| ------------------------- | ------------------------------- | --------------- |
| Linear Regression         | Baseline                        | 0.7530          |
| **Linear Regression**     | **+ Log Transformation**        | **0.7716 ✅**    |
| Linear Regression         | + Log + Feature Scaling         | 0.7716          |
| Polynomial Regression     | Baseline                        | 0.8349          |
| **Polynomial Regression** | **+ Log Transformation**        | **0.8495**    |
| Polynomial Regression     | + Log + Feature Scaling         | 0.8495          |
| **SVR**                   | **RBF + Feature Scaling**       | **0.8377**    |
| SVR                       | Polynomial (degree=2)           | 0.8347          |
| SVR                       | Polynomial (degree=3)           | 0.8235          |
| SVR                       | Polynomial (degree=4)           | 0.7935          |
| SVR                       | Linear Kernel                   | 0.7043          |
| SVR                       | Sigmoid Kernel                  | -254.12        |
| SVR                       | RBF + Log Transformation        | -351M+         |
| Decision Tree             | No Tune, no log                 | 0.6860          |
| Decision Tree             | No Tune + Log                   | -385M+         |
| Decision Tree             | Tuned + Log                     | -340M+         |
| Decision Tree             | Tuned, no log (max_depth=3)     | 0.8306          |
| Decision Tree             | Tuned, no log (max_depth=5)     | 0.8410          |
| **Decision Tree**         | **Tuned, no log (max_depth=6)** | **0.8417**    |
| Decision Tree             | Tuned, no log (max_depth=7)     | 0.8277          |

